### PR TITLE
Feature/xivy 2275 engine to target

### PR DESCRIPTION
**Pull request by ivy-ohe**

### CODE REVIEW

**Comment by ivy-ohe** [comment]
See: [XIVY-2275](https://jira.axonivy.com/jira/browse/XIVY-2275)
New parameter <engineToTarget> when set to TRUE it will clone the engine from the cache (engineDirectory) to the target folder of the maven project and use this one instead.

Cons:
- lots of IO
- reactor && engineToTarget could use up a lot of disk space (if you clone the engine in each target)
- can't share the "target" engine between projects

Pros:
- you could pre-configure the original engine and don't have to worry that it gets dirty
- each execution uses a "clean" engine

**Comment by ivy-ohe** [comment]
Build: https://jenkins.ivyteam.io/job/project-build-plugin/job/feature%252FXIVY-2275_engineToTarget/

File: src/main/java/ch/ivyteam/ivy/maven/AbstractEngineMojo.java [hunk 1]
```diff
@@ -41,6 +44,9 @@
   
   protected static final String ENGINE_DIRECTORY_PROPERTY = "ivy.engine.directory";
   
+  @Parameter(property = "project", required = true, readonly = true)
+  protected MavenProject project;
```
Comment (ivy-rew) [inline src/main/java/ch/ivyteam/ivy/maven/AbstractEngineMojo.java]: does the plain 'install-engine-mojo' > comand line usage still work ? I guess not ... project requires mvn to be executed with a project . 
But install-engine-mojo can provide one for other build pipelines than maven. think of a chef or Puppet script to setup your complete axon.ivy environment from scratch....

File: src/main/java/ch/ivyteam/ivy/maven/AbstractEngineMojo.java [hunk 2]
```diff
@@ -129,9 +142,36 @@ protected final File findMatchingEngineInCacheDirectory() throws MojoExecutionEx
...
+    
+    if (engineToTarget && project != null && engineDirToTake != null)
+    {
+      engineDirToTake = copyEngineToTarget(engineDirToTake);
+    }
+    
     return engineDirToTake;
   }
 
+  private File copyEngineToTarget(File engineDirToTake)
```
Comment (ivy-rew) [inline src/main/java/ch/ivyteam/ivy/maven/AbstractEngineMojo.java]: how it the intended behaviour in multi-module builds? will the same engine be used for all projects? or one per project? .

File: src/main/java/ch/ivyteam/ivy/maven/AbstractProjectCompileMojo.java [hunk 3]
```diff
@@ -42,9 +41,6 @@
 
 public abstract class AbstractProjectCompileMojo extends AbstractEngineMojo
 {
-  @Parameter(property = "project", required = true, readonly = true)
```
Comment (ivy-rew) [inline src/main/java/ch/ivyteam/ivy/maven/AbstractProjectCompileMojo.java]: I think the compile/test goal should not be affected by this change. the provided engine is being manipulated between the start-test-engine and stop-test-engine goal. There an isolated engine can be handy.... here for compilation I do not see much value: but lot's of potential unnecessary engine copying

Comment (ivy-rew) [reply src/main/java/ch/ivyteam/ivy/maven/AbstractProjectCompileMojo.java]: the longer I think about it : the whole flag and its processing should only be in the start/stop goals. There we could even warn users that testing against a cached engine is a bad choice (or turn the feature of in there completely...)

File: src/main/java/ch/ivyteam/ivy/maven/AbstractEngineMojo.java (hunk 1 above)
Comment (ivy-ohe) [reply src/main/java/ch/ivyteam/ivy/maven/AbstractEngineMojo.java]: I ran:
mvn com.axonivy.ivy.ci:project-build-plugin:8.0.4-SNAPSHOT:installEngine

without a project and it worked without any issues.

File: src/main/java/ch/ivyteam/ivy/maven/AbstractProjectCompileMojo.java (hunk 3 above)
Comment (ivy-rew) [reply src/main/java/ch/ivyteam/ivy/maven/AbstractProjectCompileMojo.java]: another positive thing if isolated in the integration tests goals: we can conveniently write a unit test for this case....

File: src/main/java/ch/ivyteam/ivy/maven/AbstractEngineMojo.java (hunk 2 above)
Comment (ivy-ohe) [reply src/main/java/ch/ivyteam/ivy/maven/AbstractEngineMojo.java]: This implementation would copy the engine for each project, sharing is not possible.

File: src/main/java/ch/ivyteam/ivy/maven/AbstractProjectCompileMojo.java (hunk 3 above)
Comment (ivy-ohe) [reply src/main/java/ch/ivyteam/ivy/maven/AbstractProjectCompileMojo.java]: in hindsight it does not make much sense to provide this feature for compiling...

I'll have another look at the start/stop test engine mojos

Thank you for the quick feedback

**Pull request closed without merging**